app/models/viewmodels/notificationspage/NotificationPageSingleElementFactory.scala

```diff
@@ -76,13 +77,14 @@ class NotificationPageSingleElementFactory @Inject()(
     buildForRequest(submission).copy(content = content)
   }
 
+  private val converters =
+    Set(controlResponseConverter, ersResponseConverter, emrResponseConverter, movementResponseConverter)
+
   def build(notification: NotificationFrontendModel)(implicit messages: Messages): NotificationsPageSingleElement =
-    notification.responseType match {
-      case ControlResponse        => controlResponseConverter.convert(notification)
-      case MovementTotalsResponse => movementTotalsResponseConverter.convert(notification)
-      case MovementResponse       => movementResponseConverter.convert(notification)
-      case _                      => buildForUnspecified(notification.timestampReceived)
-    }
+    converters
+      .find(_.canConvertFrom(notification))
+      .map(_.convert(notification))
+      .getOrElse(buildForUnspecified(notification.timestampReceived))
```

Comment:
I think it'll be easier to stay with pattern matching.

```
(notification.responseType, notification.messageCode) match {
    case (ControlResponse, _) => ???
    case (MovementTotalsResponse, EMR.code) => ???
    case ... etc
}
```

Reply:
Here i am not sure - Maciek's version tails better story about what actual happen in reality - and IMO it is easier to read

Reply:
Staying with pattern matching here would mean the logic leaking from converters to factory. Also, since the converters 'knows' what they can convert from, there is no point in doubling this part of the logic.